### ISSUE TITLE
Unable to convert tank data

### ISSUE THREAD

**Issue by Ormira** [SignalK/signalk-to-nmea2000#92]
I have tank data from victonr plugin in signalk but the nmea2000 plugin doesn't convert them to nmea2000 : 
![image](https://github.com/SignalK/signalk-to-nmea2000/assets/10291565/2bcb27e4-a51d-43ee-a03b-2dd3604246fc)

The plugin seems to work, I've tested outside temperature and the data are visible on my nmea network.

```
Feb 29 11:20:52 2024-02-29T11:20:52.363Z signalk-to-nmea2000 Tank Levels (127505) is enabled
Feb 29 11:20:52 2024-02-29T11:20:52.376Z signalk-to-nmea2000 Outside Temperature (130312) is enabled
Feb 29 11:20:52 2024-02-29T11:20:52.381Z signalk-to-nmea2000 keys:environment.outside.temperature
Feb 29 11:20:52 2024-02-29T11:20:52.384Z signalk-to-nmea2000 timeouts:
Feb 29 11:20:52 2024-02-29T11:20:52.394Z signalk-to-nmea2000 environment.outside.temperature undefined
Feb 29 11:20:55 signalk-server running at 0.0.0.0:3000
Feb 29 11:21:03 2024-02-29T11:21:03.651Z signalk-to-nmea2000 environment.outside.temperature:286.55499973297117
Feb 29 11:21:03 2024-02-29T11:21:03.660Z signalk-to-nmea2000 emit nmea2000JsonOut {"pgn":130312,"SID":255,"Instance":101,"Source":1,"Actual Temperature":286.55499973297117} 
```
I've also tried to input data in the data fiddler with no more result.

What can I do to check further ?

signalk version : 2.5.0
plugin version : 2.17.1

**Comment by sbender9** [comment]
Can you attach the plugin config file.  ~/.signalk/plugin-config-data/sk-to-nmea2000.json

**Comment by Ormira** [comment]
Thanks for your prompt response. Here is the config file :
[sk-to-nmea2000.json](https://github.com/SignalK/signalk-to-nmea2000/files/14450192/sk-to-nmea2000.json)

**Comment by sbender9** [comment]
You need to add tank mappings in the plugin config.

**Comment by Ormira** [comment]
Oh ! Of course. What is the syntax ? for the other like batteries or solar there is a mapper : 
![image](https://github.com/SignalK/signalk-to-nmea2000/assets/10291565/82b40d7c-54d5-4b1d-b3e7-5c46472e316c)
but not for the tanks : 
![image](https://github.com/SignalK/signalk-to-nmea2000/assets/10291565/5296e5d3-2089-4443-86db-77e46d2d4baf)

**Comment by sbender9** [comment]
Something is strange, it should be there. 

![image](https://github.com/SignalK/signalk-to-nmea2000/assets/7853683/7428e75f-c752-4f78-a507-e9a31b0b28a5)

**Comment by sbender9** [comment]
I just checked on another server of mine that doesn't have any mappings and I'm seeing the same as you there. 

Let me see if I can fix it now...

**Comment by Ormira** [comment]
Thanks.
Let me know if I can help. If it's not easy to fix, I can edit the conf file manually if I know how the data should look.

**Comment by sbender9** [comment]
ok. That happens if there is no data for any tanks. Wait for tank data to show up in the Data Browser, then go to the plugin config.

**Comment by sbender9** [comment]
And might need to refresh that page if you went there before tank data showed up. Not sure.

**Comment by Ormira** [comment]
Nothing seems to work. I've tried : 

1. waiting for tank data in the data browser, then going to the plugin config
2. desactivate the plugin, reboot the server, empty the firefox cache, waiting for tank data in the data browser then activating the plugin
3. rebooting again, empty the firefox cache, waiting for tank data in the data browser

each time the plugin config page is the same.

**Comment by sbender9** [comment]
Very strange!

Try it manually:

```
"TANKS": {
      "enabled": false,
      "resend": 5,
      "resendTime": 30,
      "tanks": [
        {
          "signalkPath": "tanks.freshWater.2",
          "instanceId": 0
        },
        {
          "signalkPath": "tanks.blackWater.1",
          "instanceId": 0
        },
        {
          "signalkPath": "tanks.fuel.3",
          "instanceId": 0
        },
        {
          "signalkPath": "tanks.lubrication.4",
          "instanceId": 0
        }
      ]
    },
```

**Comment by Ormira** [comment]
It works !
```
Feb 29 22:49:11 2024-02-29T22:49:11.481Z signalk-to-nmea2000 Tank Levels (127505) is enabled
Feb 29 22:49:11 2024-02-29T22:49:11.484Z signalk-to-nmea2000 keys:tanks.freshWater.22.currentLevel,tanks.freshWater.22.capacity
Feb 29 22:49:11 2024-02-29T22:49:11.486Z signalk-to-nmea2000 timeouts:60000,60000
Feb 29 22:49:11 2024-02-29T22:49:11.489Z signalk-to-nmea2000 tanks.freshWater.22.currentLevel undefined
Feb 29 22:49:11 2024-02-29T22:49:11.493Z signalk-to-nmea2000 tanks.freshWater.22.capacity undefined
Feb 29 22:49:11 2024-02-29T22:49:11.505Z signalk-to-nmea2000 keys:tanks.fuel.23.currentLevel,tanks.fuel.23.capacity
Feb 29 22:49:11 2024-02-29T22:49:11.507Z signalk-to-nmea2000 timeouts:60000,60000
Feb 29 22:49:11 2024-02-29T22:49:11.510Z signalk-to-nmea2000 tanks.fuel.23.currentLevel undefined
Feb 29 22:49:11 2024-02-29T22:49:11.514Z signalk-to-nmea2000 tanks.fuel.23.capacity undefined
Feb 29 22:49:11 POST /skServer/plugins/sk-to-nmea2000/config 200 15.128 ms - 45
Feb 29 22:49:11 GET /skServer/plugins 200 128.866 ms - -
Feb 29 22:49:15 GET /skServer/debugKeys 200 13.740 ms - -
Feb 29 22:49:23 2024-02-29T22:49:23.726Z signalk-to-nmea2000 tanks.fuel.23.currentLevel,tanks.fuel.23.capacity:0.92,0.12999999523162842
Feb 29 22:49:23 2024-02-29T22:49:23.730Z signalk-to-nmea2000 emit nmea2000JsonOut {"pgn":127505,"Instance":0,"Type":"Fuel","Level":92,"Capacity":129.99999523162842}
Feb 29 22:49:23 2024-02-29T22:49:23.736Z signalk-to-nmea2000 tanks.freshWater.22.currentLevel,tanks.freshWater.22.capacity:0.08,0.33000001311302185
Feb 29 22:49:23 2024-02-29T22:49:23.739Z signalk-to-nmea2000 emit nmea2000JsonOut {"pgn":127505,"Instance":0,"Type":"Water","Level":8,"Capacity":330.00001311302185}
```

but still no mapping in the plugin config page.

If you want to dig further on the issue I'll be happy to help, otherwise I consider the problem solved for me.

Thanks again !

**Comment by albuchan** [comment]
I was following this issue through. I have also an issue with tanks giving no output on 127505

In my case I have one fuel tank. I am generating the values from node red and updating the paths in sk.
Here you can see the paths in the databrowser.
![image](https://github.com/SignalK/signalk-to-nmea2000/assets/74618562/691a8169-cab0-492e-a349-8d6bacca37e6)

Looking at the log, I think the plugin only uses these keys  keys: tanks.fuel.0.currentLevel,tanks.fuel.0.capacity
But I had generated other keys for currentVolume, name and type as I saw them in the 127505 msg spec. 
I do not have the "remaining" key generated by victron shown by ormira.

Note. I currently have tanks.fuel.0.type="diesel" as this is what I saw as a defined value in the Signalk Path spec.
However I have changed this value to both "fuel" or "Fuel" after seeing the working example output from ormira above.
But still doesn't work for me.

This is from the initial debug log entries on enabling signalk-to-nmea2000.
Mar 10 13:21:49 2024-03-10T12:21:49.732Z signalk-to-nmea2000 Tank Levels (127505) is enabled
Mar 10 13:21:49 2024-03-10T12:21:49.733Z signalk-to-nmea2000 keys:tanks.fuel.0.currentLevel,tanks.fuel.0.capacity
Mar 10 13:21:49 2024-03-10T12:21:49.733Z signalk-to-nmea2000 timeouts:60000,60000
Mar 10 13:21:49 2024-03-10T12:21:49.733Z signalk-to-nmea2000 tanks.fuel.0.currentLevel undefined
Mar 10 13:21:49 2024-03-10T12:21:49.733Z signalk-to-nmea2000 tanks.fuel.0.capacity undefined

Although both keys immediately log as "undefined", this is also the case for other plugin pgns and their keys I have enabled, 
but then those other PGN parsers seem to find the values and transmit the values.
I have no issue in the config. fuel.0 appears in the dropdown and I have selected it, submited, and restarted both plugin and sk.
![image](https://github.com/SignalK/signalk-to-nmea2000/assets/74618562/7db56b4f-419f-46f7-81f6-8eb1a9595051)
Also looks the same in the .json config file so I have not copied it here.

So my issue is that unlike Ormira above..having got a definition set, and providing the key data. I simply don't get the corresponding outputs as copied here. No PGN 127505 at all no matter how long I run the plugin to debug output.
***********     From Ormira ***************
Feb 29 22:49:23 2024-02-29T22:49:23.726Z signalk-to-nmea2000 tanks.fuel.23.currentLevel,tanks.fuel.23.capacity:0.92,0.12999999523162842
Feb 29 22:49:23 2024-02-29T22:49:23.730Z signalk-to-nmea2000 emit nmea2000JsonOut {"pgn":127505,"Instance":0,"Type":"Fuel","Level":92,"Capacity":129.99999523162842}
*********************************************
Version
2.17.1

SK version is 2.5.0  ( I am also getting failures upgrading to 2.6.0).

**Comment by albuchan** [comment]
Update to this. I disabled all the other PGN's and 127505 did start to transmit even got recognized on my raymarine axiom although tank was now defined it didn't show capacity. However, as I started to enable Engine Parameters (127489,127488), and my 3 defined Batteries (127506 & 127508)  the 127505 is no longer transmitted. Could this be a timing issue ( currently I have engine and battery @ 30,30 seconds retransmit)?
